go/src/lib.rs

```diff
@@ -1,9 +1,171 @@
+// TODO: Change message type when implementing command logic
+// TODO: Consider using a single response callback instead of success and failure callbacks
+pub type SuccessCallback = unsafe extern "C" fn(index_ptr: usize, message: *const c_char) -> ();
+
+/// Failure callback that is called when a Redis command fails.
+///
+/// `index_ptr` is a baton-pass back to the caller language to uniquely identify the promise.
+/// `error_message` is the error message returned by Redis for the failed command. The lifetime of `error_message` is until the callback returns control back to the caller. The callee cannot use the data after the callback returns.
+/// `error_type` is the type of error returned by glide-core, depending on the `RedisError` returned.
+pub type FailureCallback = unsafe extern "C" fn(
+    index_ptr: usize,
+    error_message: *const c_char,
+    error_type: RequestErrorType,
+) -> ();
+
+/// The connection response.
+///
+/// It contains either a connection or an error. It is represented as a struct instead of a union for ease of use in the wrapper language.
+///
+/// The struct is freed using `free_connection_response` to avoid memory leaks. This is achieved by having the external caller invoke this function.
+#[repr(C)]
+pub struct ConnectionResponse {
+    conn_ptr: *const c_void,
+    connection_error_message: *const c_char,
+}
+
+/// The glide client.
+// TODO: Remove allow(dead_code) once connection logic is implemented
+#[allow(dead_code)]
+pub struct Client {
+    client: GlideClient,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+    runtime: Runtime,
+}
+
+fn create_client_internal(
+    connection_request_bytes: &[u8],
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> Result<Client, String> {
+    let request = connection_request::ConnectionRequest::parse_from_bytes(connection_request_bytes)
+        .map_err(|err| err.to_string())?;
+    // TODO: optimize this (e.g. by pinning each go thread to a rust thread)
+    let runtime = Builder::new_multi_thread()
+        .enable_all()
+        .worker_threads(1)
+        .thread_name("GLIDE for Redis Go thread")
+        .build()
+        .map_err(|err| {
+            let redis_error = err.into();
+            errors::error_message(&redis_error)
+        })?;
+    let _runtime_handle = runtime.enter();
+    let client = runtime
+        .block_on(GlideClient::new(ConnectionRequest::from(request)))
+        .map_err(|err| err.to_string())?;
+    Ok(Client {
+        client,
+        success_callback,
+        failure_callback,
+        runtime,
+    })
+}
+
+/// Creates a new client with the given configuration. The success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns. All callbacks should be offloaded to separate threads in order not to exhaust the client's thread pool.
```

Comment:
> he success callback needs to copy the given string synchronously, since it will be dropped by Rust once the callback returns

these details are also true for the failure callback, and in case should be a part of the callback docs, not here.

Reply:
Consider adding links to the callback docs

Reply:
I already moved the documentation for the callbacks in the latest revision.